dbms/src/Flash/Mpp/ReceivedMessageQueue.h

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/FailPoint.h>
+#include <Common/LooseBoundedMPMCQueue.h>
+#include <Flash/Mpp/GRPCReceiveQueue.h>
+#include <Flash/Mpp/TrackedMppDataPacket.h>
+
+#include <memory>
+
+namespace DB
+{
+enum class ReceiverMode
+{
+    Local = 0,
+    Sync,
+    Async
+};
+
+struct ReceivedMessage
+{
+    size_t source_index;
+    String req_info;
+    // shared_ptr<const MPPDataPacket> is copied to make sure error_ptr, resp_ptr and chunks are valid.
+    const std::shared_ptr<DB::TrackedMppDataPacket> packet;
+    const mpp::Error * error_ptr;
+    const String * resp_ptr;
+    std::vector<const String *> chunks;
+    /// used for fine grained shuffle
+    std::vector<std::vector<const String *>> fine_grained_chunks;
+    std::shared_ptr<std::atomic<size_t>> remaining_consumers;
```

Comment:
Add some comments here?
When non-fine-grained shuffle, remaining_consumers is null ...

Reply:
Done